uttut/pipeline/ops/span_subwords.py

```diff
@@ -0,0 +1,173 @@
+        output_sequence = lst2lst.apply(input_sequence, forward_replacement_group)
+        inverse_replacement_group = lst2lst.inverse(input_sequence, forward_replacement_group)
+
+        updated_labels = propagate_by_replacement_group(
+            labels, forward_replacement_group, self._forward_reduce_func)
+
+        realigner = self._realigner_class(
+            edit=inverse_replacement_group,
+            input_length=len(output_sequence),
+            output_length=len(input_sequence),
+        )
+
+        return output_sequence, updated_labels, realigner
+
+    def _gen_forward_replacement_group(self, input_lst: List[str]) -> ReplacementGroup:
+
+        replacement_group = ReplacementGroup()
+        for i, token in enumerate(input_lst):
+            subtokens = span_subwords(
+                word=token,
+                unk_token=self.unk_token,
+                vocab=self.vocab,
+                max_input_chars_per_word=self.maxlen_per_token,
+            )
+            if subtokens != [token]:
+                replacement_group.add(start=i, end=i + 1, new_value=subtokens,
+                                      annotation='span-subwords')
+        replacement_group.done()
+
+        return replacement_group
+
+    def _forward_reduce_func(self, labels: List[int], output_size: int):
+        return get_most_common(labels, output_size)
+
+
+class WordPieceRealigner(Realigner):
+
+    def _backward_transduce_func(self, labels: List[int], output_size: int) -> List[int]:
+        return get_most_common_except_not_entity(labels, output_size)
+
+    def _realign_labels(self, labels: List[int]) -> List[int]:
+        return propagate_by_replacement_group(labels, self._edit, self._backward_transduce_func)
+
+
+def span_subwords(
+        word: str,
+        unk_token: str,
+        vocab: Dict[str, int],
+        max_input_chars_per_word: int,
+    ) -> List[str]:
+
+    """
+    Tokenize a word into several subwords.
+
+    This code is copied from Bert tokenization.py.
+
+    Requirement: The input word should not contain whitespaces and accent tokens.
+
+    This uses a greedy longest-match-first algorithm to perform tokenization
+    using the given vocabulary.
+
+    The default arguments in BERT are
+    unk_token = '[UNK]'
+    max_input_chars_per_word = 200
+
+    E.g.
+    >>> span_subwords(
+        word="unaffable",
+        vocab={"un": 0, "##aff": 1, "##able": 2},
+        max_input_chars_per_word=200,
+    )
+    ["un", "##aff", "##able"]
+
+    """
+
+    # length of word is more than expected -> return [unk_token]
+    if len(word) > max_input_chars_per_word:
+        return [unk_token]
+
+    # subword matching
+    chars = list(word)
+    is_bad = False
+    start = 0
+    sub_tokens = []
+    while start < len(chars):
+        end = len(chars)
+        cur_substr = None
+        while start < end:
+            substr = "".join(chars[start: end])
```

Comment:
`chars = list(word) ` and `substr = "".join(chars[start:end])`

can just be `substr = word[start:end]`

Reply:
This code is completely copied from BERT. Should I modify it?